lib/mail/message.rb

```diff
@@ -1994,14 +1992,44 @@ def text?
     # Additionally, I allow for the case where someone might have put whitespace
     # on the "gap line"
     def parse_message
-      header_part, body_part = raw_source.lstrip.split(HEADER_SEPARATOR, 2)
-      self.header = header_part
-      self.body   = body_part
+      stream = StringIO.new(raw_source)
+      self.header = parse_headers(stream)
+      self.body = stream.read
+    end
+
+    # Parses the headers from the input stream,
+    # leaving the position at the start of the body.
+    def parse_headers(stream)
+      buffer = ''
+      first = true
+      loop do
+        line = stream.gets
+        break if line.nil?
+        if line.strip.empty?
+          next if first && headers_required?
+          break
+        end
+        line.rstrip!
+        stream.rewind && break if !headers_required? && first && !(line =~ HEADER_LINE)
+        if line.start_with?(' ') || line.start_with?("\t")
+          line = line.strip if first
+          buffer += line
+        else
+          buffer += CRLF unless first
+          buffer += line
+        end
+        first = false
+      end
+      buffer
+    end
```

Comment:
This is difficult to check for correctness. Please comment the the implementation to indicate intent and logic flow.